automerge-backend/src/protocol.rs

```diff
@@ -87,63 +87,58 @@ impl ActorID {
 pub struct Clock(pub HashMap<ActorID, u32>);
 
 impl Clock {
-    pub fn new(actor_id: &ActorID, seq: u32) -> Clock {
-        let mut map = HashMap::new();
-        map.insert(actor_id.clone(), seq);
-        Clock(map)
-    }
-
     pub fn empty() -> Clock {
         Clock(HashMap::new())
     }
 
-    pub fn with_dependency(&self, actor_id: &ActorID, new_seq: u32) -> Clock {
-        let mut result = self.0.clone();
-        result.insert(actor_id.clone(), new_seq);
-        Clock(result)
+    pub fn with(&self, actor_id: &ActorID, seq: u32) -> Clock {
+        let mut result = self.clone();
+        result.set(actor_id, max(seq, self.get(actor_id)));
+        result
     }
 
-    pub fn upper_bound(&self, other: &Clock) -> Clock {
-        let mut result: HashMap<ActorID, u32> = HashMap::new();
-        self.0.iter().for_each(|(actor_id, seq)| {
-            result.insert(
-                actor_id.clone(),
-                max(*seq, *other.0.get(actor_id).unwrap_or(&(0 as u32))),
-            );
-        });
-        other.0.iter().for_each(|(actor_id, seq)| {
-            result.insert(
-                actor_id.clone(),
-                max(*seq, *self.0.get(actor_id).unwrap_or(&(0 as u32))),
-            );
+    pub fn merge(&mut self, other: &Clock) {
+        other.into_iter().for_each(|(actor_id, seq)| {
+            self.set(actor_id, max(*seq, self.get(actor_id)));
         });
-        Clock(result)
     }
 
-    pub fn is_before_or_concurrent_with(&self, other: &Clock) -> bool {
-        other
-            .0
-            .iter()
-            .all(|(actor_id, seq)| self.0.get(actor_id).unwrap_or(&0) >= seq)
+    pub fn union(&self, other: &Clock) -> Clock {
```

Comment:
I'm not sure I like naming this `union`, the problem is that there are many possible unions for a clock as you have to decide what to do when an actor exists in both clocks with different values. Mostly what we do in Automerge is take the maximum, but you can also take the minimum. The "upper bound" name was from the idea of vector clocks as a lattice, but we could use some other name if it's not very clear, maybe `max`?

Reply:
In the automerge context clocks only go up - never down.  And at least for me when I think about the union of two clocks its natural to think of it as the upper bound of both.  I get that if this were a general purpose library union would make less sense... `max` also feels good to me.  I only prefer `union` b/c it seems very clear to me the result is a new clock and not a mutation of the old clock. Your call tho - I can go either way

Reply:
Let's go with `max`, I agree that we never need anything other than `max` but it's clearer to me when reading the code what the semantics are.